File: cpp/test/test_rand2.cpp
#include <stdio.h>
#include "TRTCContext.h"
#include "DVVector.h"
#include "DVRNG.h"

int main()
{
	DVRNG& rng = DVRNG::singlton();

	TRTC_For ker_init({ "rng", "vec_states" }, "idx",
		"    rng.state_init(1234, idx, 0, vec_states[idx]);\n"
	);

	TRTC_For ker_call({ "vec_states", "vec_rnd" }, "idx",
		"    vec_rnd[idx]=(float)vec_states[idx].rand01();\n"
	);

	DVVector d_vec_states("RNGState", 1024);
	DVVector d_vec_rnd("float", 1024);
	{
		const DeviceViewable* args[] = { &rng, &d_vec_states };
		ker_init.launch_n(1024, args);
	}
	{
		const DeviceViewable* args[] = { &d_vec_states, &d_vec_rnd };
		ker_call.launch_n(1024, args);
	}

	float h_vec_rnd[1024];
	d_vec_rnd.to_host(h_vec_rnd);

	for (int i = 0; i < 1024; i++)
		printf("%f ", h_vec_rnd[i]);

	printf("\n");

	return 0;
}

File: python/test/test_rand2.py
# example showing how to separate init and call

import ThrustRTC as trtc
import CURandRTC as rndrtc

rng = rndrtc.DVRNG()

ker_init = trtc.For(['rng','states'], 'idx',
	'''
	rng.state_init(1234, idx, 0, states[idx]);
	''')

ker_call = trtc.For(['states', 'vec_rnd'], 'idx',
	'''
	vec_rnd[idx]=(float)states[idx].rand01();
	'''
	)

rng_states = trtc.device_vector('RNGState', 1024)
d_vec_rnd = trtc.device_vector('float', 1024)

ker_init.launch_n(1024, [rng, rng_states])
ker_call.launch_n(1024, [rng_states, d_vec_rnd])
print (d_vec_rnd.to_host())
